core/src/main/java/org/elasticsearch/index/percolator/QueryMetadataService.java

```diff
@@ -0,0 +1,192 @@
+            Term longestTerm = terms[0];
+            for (Term term : terms) {
+                if (longestTerm.bytes().length < term.bytes().length) {
+                    longestTerm = term;
+                }
+            }
+            return Collections.singletonList(longestTerm);
+        } else if (query instanceof BooleanQuery) {
+            List<BooleanClause> clauses = ((BooleanQuery) query).clauses();
+            boolean hasRequiredClauses = false;
+            for (BooleanClause clause : clauses) {
+                if (clause.isRequired()) {
+                    hasRequiredClauses = true;
+                    break;
+                }
+            }
+            if (hasRequiredClauses) {
+                List<Term> bestClause = null;
+                for (BooleanClause clause : clauses) {
+                    if (clause.isProhibited()) {
+                        // we don't need to remember the things that do *not* match...
+                        continue;
+                    }
+                    if (clause.getOccur() == BooleanClause.Occur.SHOULD) {
+                        // this bq has must clauses, so we don't need to remember should clauses,
+                        // since they are completely optional.
+                        continue;
+                    }
+
+                    List<Term> temp = extractQueryMetadata(clause.getQuery());
+                    bestClause = selectTermsListWithHighestSumOfTermLength(temp, bestClause);
+                }
+                if (bestClause != null) {
+                    return bestClause;
+                } else {
+                    return Collections.emptyList();
+                }
+            } else {
+                List<Term> terms = new ArrayList<>();
+                for (BooleanClause clause : clauses) {
+                    if (clause.isProhibited()) {
+                        // we don't need to remember the things that do *not* match...
+                        continue;
+                    }
+                    terms.addAll(extractQueryMetadata(clause.getQuery()));
+                }
+                return terms;
+            }
+        } else if (query instanceof ConstantScoreQuery) {
+            Query wrappedQuery = ((ConstantScoreQuery) query).getQuery();
+            return extractQueryMetadata(wrappedQuery);
+        } else if (query instanceof BoostQuery) {
+            Query wrappedQuery = ((BoostQuery) query).getQuery();
+            return extractQueryMetadata(wrappedQuery);
+        } else {
+            throw new IllegalArgumentException("unsupported query");
+        }
+    }
+
+    List<Term> selectTermsListWithHighestSumOfTermLength(List<Term> terms1, List<Term> terms2) {
+        if (terms1 == null) {
+            return terms2;
+        } else if (terms2 == null) {
+            return terms1;
+        } else {
+            int terms1SumTermLength = computeSumOfTermLength(terms1);
```

Comment:
maybe an average would make more sense? otherwise this would make clauses that have lots of terms too likely to be picked? eg. if the query was `foo AND (b OR a OR r OR z)` then we would index `(b,a,r,z)` while `foo` would probably be a better decision? Or maybe the clause whose shortest term is longer? (sorry I'm thinking out loud)